File: gen_server_tutorial/src/gen_server_by_py/counter_server.py
"""
counter_server.py  –  A Python gen_server callback module (counter example).

Architecture Mapping (from gen_server_architecture.tex):

  This file implements BOTH:
    - Left-Top:    Client API (thin wrappers around gen_server.call/cast/stop)
    - Left-Bottom: Callback Module (user-written callbacks)

  Equivalent to counter_server.erl:
    -module(counter_server).
    -behaviour(gen_server).

  After start_link(CounterServer, ...):
    fun counter_server:init/1
    fun counter_server:handle_call/3
    fun counter_server:handle_cast/2
    fun counter_server:handle_info/2
    fun counter_server:handle_continue/2
    fun counter_server:terminate/2

  The ?MODULE info is cached in the GenServerProcess (≈ #server_data{}).
"""

from __future__ import annotations

import threading
from typing import Any, Tuple

import gen_server
from gen_server import GenServerBehaviour, GenServerProcess


# ============================================================
# Callback Module (bottom-left quadrant)
# Equivalent to:
#   -module(counter_server).
#   -behaviour(gen_server).
# ============================================================

class CounterServer(GenServerBehaviour):
    """
    A minimal gen_server callback module — counter example.

    State: {'count': int, 'history': list}

    Callbacks:
        init/1            → initialize counter
        handle_call/3     → increment, decrement, get_count, slow_op
        handle_cast/2     → reset, crash
        handle_info/2     → custom messages
        handle_continue/2 → post_init
        terminate/2       → cleanup
    """

    # ---- init/1 ----
    def init(self, args: Any) -> Tuple:
        """
        Initialize the counter state.

        Equivalent to:
            init(InitCount) when is_integer(InitCount) ->
                process_flag(trap_exit, true),
                {ok, #{count => InitCount, history => []},
                 {continue, post_init}}.
        """
        init_count = args if isinstance(args, int) else 0
        print(f"[counter] init: count={init_count}, "
              f"tid={threading.current_thread().ident}")

        state = {'count': init_count, 'history': []}
        return ('ok', state, {'continue': 'post_init'})

    # ---- handle_continue/2 ----
    def handle_continue(self, continue_arg: Any, state: Any) -> Tuple:
        """
        Post-initialization hook.

        Equivalent to:
            handle_continue(post_init, State) ->
                io:format("[counter] post_init complete~n"),
                {noreply, State}.
        """
        if continue_arg == 'post_init':
            print("[counter] post_init complete")
        return ('noreply', state)

    # ---- handle_call/3 ----
    def handle_call(self, request: Any, from_ref: Any, state: Any) -> Tuple:
        """
        Handle synchronous requests.

        Equivalent to:
            handle_call(increment, _From, #{count := C} = State) -> ...
            handle_call(decrement, _From, #{count := C} = State) -> ...
            handle_call(get_count, _From, #{count := C} = State) -> ...
            handle_call({slow_op, Duration}, _From, State) -> ...
        """
        if request == 'increment':
            new_count = state['count'] + 1
            state['count'] = new_count
            state['history'].append(('increment', new_count))
            return ('reply', new_count, state)

        elif request == 'decrement':
            new_count = state['count'] - 1
            state['count'] = new_count
            state['history'].append(('decrement', new_count))
            return ('reply', new_count, state)

        elif request == 'get_count':
            return ('reply', state['count'], state)

        elif isinstance(request, tuple) and request[0] == 'slow_op':
            import time
            duration = request[1]
            time.sleep(duration)
            return ('reply', ('ok', 'done_after', duration), state)

        else:
            return ('reply', ('error', 'unknown_request'), state)

    # ---- handle_cast/2 ----
    def handle_cast(self, msg: Any, state: Any) -> Tuple:
        """
        Handle asynchronous messages.

        Equivalent to:
            handle_cast(reset, State) -> {noreply, State#{count := 0}}.
            handle_cast(crash, _State) -> error(deliberate_crash).
        """
        if msg == 'reset':
            print("[counter] reset to 0")
            state['count'] = 0
            state['history'] = []
            return ('noreply', state)

        elif msg == 'crash':
            raise RuntimeError("deliberate_crash")

        else:
            return ('noreply', state)

    # ---- handle_info/2 ----
    def handle_info(self, info: Any, state: Any) -> Tuple:
        """
        Handle raw messages.

        Equivalent to:
            handle_info({custom_msg, Payload}, State) -> ...
            handle_info(Info, State) -> ...
        """
        if isinstance(info, tuple) and len(info) == 2 and info[0] == 'custom_msg':
            print(f"[counter] received custom_msg: {info[1]!r}")
            return ('noreply', state)

        print(f"[counter] unexpected info: {info!r}")
        return ('noreply', state)

    # ---- terminate/2 ----
    def terminate(self, reason: Any, state: Any) -> None:
        """
        Cleanup on shutdown.

        Equivalent to:
            terminate(Reason, #{count := C}) ->
                io:format("[counter] terminating: reason=~p, final_count=~p~n",
                          [Reason, C]).
        """
        print(f"[counter] terminating: reason={reason!r}, "
              f"final_count={state.get('count', '?')}")


# ============================================================
# Client API (top-left quadrant)
# Thin wrappers around gen_server library functions.
#
# Equivalent to:
#   start_link() -> gen_server:start_link(?MODULE, 0, []).
#   increment(Server) -> gen_server:call(Server, increment).
#   ...
# ============================================================

def start_link(init_count: int = 0) -> GenServerProcess:
    """
    Start a linked counter server.

    Equivalent to:
        start_link() -> gen_server:start_link(?MODULE, 0, []).
        start_link(InitCount) -> gen_server:start_link(?MODULE, InitCount, []).

    Arrow ①: Pass module name (?MODULE = CounterServer) → bind callbacks
    """
    return gen_server.start_link(CounterServer, init_count)


def start(init_count: int = 0) -> GenServerProcess:
    """Start a standalone counter server."""
    return gen_server.start(CounterServer, init_count)


def increment(server: GenServerProcess) -> int:
    """
    Increment the counter.

    Equivalent to:
        increment(Server) -> gen_server:call(Server, increment).
    """
    return gen_server.call(server, 'increment')


def decrement(server: GenServerProcess) -> int:
    """
    Decrement the counter.

    Equivalent to:
        decrement(Server) -> gen_server:call(Server, decrement).
    """
    return gen_server.call(server, 'decrement')


def get_count(server: GenServerProcess) -> int:
    """
    Get the current count.

    Equivalent to:
        get_count(Server) -> gen_server:call(Server, get_count).
    """
    return gen_server.call(server, 'get_count')


def reset(server: GenServerProcess) -> str:
    """
    Reset the counter to 0.

    Equivalent to:
        reset(Server) -> gen_server:cast(Server, reset).
    """
    return gen_server.cast(server, 'reset')


def slow_operation(server: GenServerProcess, duration: float) -> Any:
    """
    Perform a slow operation (blocks for duration seconds).

    Equivalent to:
        slow_operation(Server, Duration) ->
            gen_server:call(Server, {slow_op, Duration}).
    """
    return gen_server.call(server, ('slow_op', duration), timeout=duration + 5)


def crash(server: GenServerProcess) -> str:
    """
    Deliberately crash the server.

    Equivalent to:
        crash(Server) -> gen_server:cast(Server, crash).
    """
    return gen_server.cast(server, 'crash')

File: gen_server_tutorial/src/counter_server.erl
%%%-------------------------------------------------------------------
%%% @doc counter_server - A minimal gen_server callback module.
%%%
%%% This module corresponds to the "Callback Module (Your Code)"
%%% quadrant (bottom-right) in the gen_server architecture diagram
%%% described in gen_server_api_guide.tex (Section 1.2).
%%%
%%% Architecture Mapping:
%%%
%%%   +---------------------+-----------------------------+
%%%   | Client API          | Generic Server              |
%%%   | (gen_server:call,   | (gen_server.erl)            |
%%%   |  cast, stop, ...)   | message loop, routing, ...  |
%%%   +---------------------+-----------------------------+
%%%   | User / Developer    | Callback Module  <-- HERE   |
%%%   | writes this module  | init/1, handle_call/3, ...  |
%%%   +---------------------+-----------------------------+
%%%
%%% The dashed arrow from "User / Developer" to "Callback Module"
%%% represents the function registration process: the module name
%%% (an atom) is passed as a parameter to gen_server:start_link/3,4,
%%% enabling late binding via functional programming.
%%%
%%% @end
%%%-------------------------------------------------------------------
-module(counter_server).
-behaviour(gen_server).

%% Client API
-export([start_link/0, start_link/1, start/0, start/1]).
-export([increment/1, decrement/1, get_count/1, reset/1]).
-export([slow_operation/2, crash/1]).
-export([demo/0]).

%% gen_server callbacks
-export([init/1, handle_call/3, handle_cast/2,
         handle_info/2, handle_continue/2,
         terminate/2, code_change/3, format_status/2]).

%%% ============================================================
%%% Client API (thin wrappers around gen_server library functions)
%%% ============================================================

start_link() ->
    gen_server:start_link(?MODULE, 0, []).

start_link(InitCount) ->
    gen_server:start_link(?MODULE, InitCount, []).

start() ->
    gen_server:start(?MODULE, 0, []).

start(InitCount) ->
    gen_server:start(?MODULE, InitCount, []).

increment(Server) ->
    gen_server:call(Server, increment).

decrement(Server) ->
    gen_server:call(Server, decrement).

get_count(Server) ->
    gen_server:call(Server, get_count).

reset(Server) ->
    gen_server:cast(Server, reset).

slow_operation(Server, Duration) ->
    gen_server:call(Server, {slow_op, Duration}).

crash(Server) ->
    gen_server:cast(Server, crash).

%%% ============================================================
%%% gen_server callbacks
%%% ============================================================

init(InitCount) when is_integer(InitCount) ->
    process_flag(trap_exit, true),
    io:format("[counter] init: count=~p, pid=~p~n",
              [InitCount, self()]),
    {ok, #{count => InitCount, history => []},
     {continue, post_init}};
init(_) ->
    {stop, {error, bad_init_arg}}.

handle_continue(post_init, State) ->
    io:format("[counter] post_init complete~n"),
    {noreply, State}.

handle_call(increment, _From, #{count := C} = State) ->
    NewCount = C + 1,
    {reply, NewCount, State#{count := NewCount,
        history := [{increment, NewCount} |
                    maps:get(history, State)]}};

handle_call(decrement, _From, #{count := C} = State) ->
    NewCount = C - 1,
    {reply, NewCount, State#{count := NewCount,
        history := [{decrement, NewCount} |
                    maps:get(history, State)]}};

handle_call(get_count, _From, #{count := C} = State) ->
    {reply, C, State};

handle_call({slow_op, Duration}, _From, State) ->
    timer:sleep(Duration),
    {reply, {ok, done_after, Duration}, State};

handle_call(_Request, _From, State) ->
    {reply, {error, unknown_request}, State}.

handle_cast(reset, State) ->
    io:format("[counter] reset to 0~n"),
    {noreply, State#{count := 0, history := []}};

handle_cast(crash, _State) ->
    error(deliberate_crash);

handle_cast(_Msg, State) ->
    {noreply, State}.

handle_info(timeout, State) ->
    io:format("[counter] timeout fired~n"),
    {noreply, State};

handle_info({custom_msg, Payload}, State) ->
    io:format("[counter] received custom_msg: ~p~n", [Payload]),
    {noreply, State};

handle_info({'EXIT', Pid, Reason}, State) ->
    io:format("[counter] linked process ~p exited: ~p~n",
              [Pid, Reason]),
    {noreply, State};

handle_info(Info, State) ->
    io:format("[counter] unexpected info: ~p~n", [Info]),
    {noreply, State}.

terminate(Reason, #{count := C}) ->
    io:format("[counter] terminating: reason=~p, "
              "final_count=~p~n", [Reason, C]),
    ok.

code_change(_OldVsn, State, _Extra) ->
    {ok, State}.

format_status(_Opt, [_PDict, #{count := C, history := H}]) ->
    [{data, [{"State",
              #{count => C,
                history_length => length(H)}}]}].

%%% ============================================================
%%% Demo
%%% ============================================================

%% @doc demo/0 - Demonstrates basic gen_server usage via counter_server.
demo() ->
    io:format("~n=== counter_server demo ===~n~n"),

    {ok, Pid} = start_link(0),
    io:format("Started: ~p~n", [Pid]),

    io:format("increment: ~p~n", [increment(Pid)]),
    io:format("increment: ~p~n", [increment(Pid)]),
    io:format("increment: ~p~n", [increment(Pid)]),
    io:format("get_count: ~p~n", [get_count(Pid)]),
    io:format("decrement: ~p~n", [decrement(Pid)]),
    io:format("get_count: ~p~n", [get_count(Pid)]),

    reset(Pid),
    timer:sleep(10),
    io:format("after reset, get_count: ~p~n", [get_count(Pid)]),

    io:format("slow_operation(100ms): ~p~n", [slow_operation(Pid, 100)]),

    gen_server:stop(Pid),
    timer:sleep(50),
    io:format("~ncounter_server demo completed.~n"),
    ok.
